Latest DataFile Only

elipe17 opened the pull request:
# Summary of Changes
- Created new celery tasks to handle individual file previous version deletion and all file previous version deletion
- Add management command for admin to execute clean task manually on a database
- Previous version cleanup for a particular file only happens on a submission not a reparse

Pull request closes #3240 

## How to Test

1. Checkout `develop`
2. Submit files, some with multiple versions and some without
3. Checkout this branch
4. Run the new management command and verify only the newest files are maintained in the DB
5. If you want to test a much larger dataset, reach out to me. I have a DB backup with ~1200 files and tens of millions of records.
6. Now you can submit multiple versions of a file and verify only it's latest version is kept in the DB
7. You should also verify that the files remain in S3 despite being deleted from the DB

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] after the nightly db backup task runs, the db will be checked for records associated with files that are not the most recent versions. and these older versions of records will be deleted from the db. all the number of records deleted, by table, will be recorded in DAC logentries. 
- [x] only latest versions of records are available to query
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [x] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Added new task to asynchroneously delete old versions of a datafile submission
- Calling task in parse routine

elipe17 committed:
- Rename task file
- Add management command to remove prev versions of datafiles that already exist in the database

elipe17 committed:
Merge branch 'develop' into 3240-latest-df-only

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3240-latest-df-only

elipe17 committed:
- Updated to use raw deletes to save memory and go faster

elipe17 committed:
Merge branch '3240-latest-df-only' of https://github.com/raft-tech/TANF-app into 3240-latest-df-only

elipe17 committed:
- Move system user to function scope
- Only delete if ids > 0

elipe17 committed:
- linting

elipe17 commented on tdrs-backend/tdpservice/scheduling/management/commands/remove_all_old_df_versions.py:
```diff
@@ -0,0 +1,19 @@
+    def handle(self, *args, **options):
+        """Remove every previous version."""
+        logger.info("Queueing task to remove all old versions of every datafile.")
+        remove_all_old_versions.delay()
```
Should we always force a DB backup for this?

ADPennington replied:
@elipe17 is a db backup actually happening here? one idea to consider is deleting the records associated with the older versions on a schedule, maybe after the nightly backup?

elipe17 replied:
There is no DB backup happening before this command executes. Since it is a command we could manually run the backup before hand to avoid introducing new celery tasks for testing. Or I could call the backup command from this command. With the impending disk constraints on our apps/backups I opt for putting the onus on the caller of this command to make the backup. Let me know your thoughts

ADPennington replied:
per async with @elipe17 cc: @lhuxraft:

- rather than deleting old records at the point of submission, moving this to a task to be run nightly after the db backup task. this enables us 24 hours to leverage the backup if we need to restore anything that was wiped out during cleanup. 
- can also be triggered from DAC
- this also has the benefit of consolidating the cleanup task in the logs (as opposed to having this cleanup dispersed throughout submissions.

elipe17 commented on tdrs-backend/tdpservice/scheduling/parser_task.py:
```diff
@@ -79,6 +80,8 @@ def parse(data_file_id, reparse_id=None):
 
             send_data_submitted_email(dfs, recipients)
+            remove_old_versions.delay(data_file_id=data_file_id,
+                                      data_file_version=data_file.version)
```
Ask Alex about celery task vs in band with parsing task

elipe17 replied:
@ADPennington when you get to this PR, let me know what you think about this. Running the cleanup as a separate task increases parallelism and load distribution, but introduces a potential delay for users. That is, they might see their previous submission history for some time before it is cleaned up. If you don't like that delay we can call this as a function instead of a celery task.

ADPennington replied:
@elipe17 i haven't started review just yet, but have a clarifying question re: scope -- does this effectively delete submission history? or just the db records associated with previous version? 

Asking because I realize there are a couple of data retention conversations happening re:
- how long to retain previous versions of files 
- how long to retain previous versions of records in the db (i originally scoped this ticket to focus on this).

elipe17 replied:
As it stands right now, this deletes everything associated to the file. Records, errors, summary, datafile object. The actual file is still preserved in S3

ADPennington replied:
Ooo okay; thank you for confirming. it's possible that I missed this scope during discussion about this ticket. there are important implications re: discarding submission history, since STTs use it for their internal audits. happy to discuss more async.

elipe17 replied:
Sure thing, let me know what we want/need to keep around of old submissions and I can make the updates.

elipe17 replied:
@ADPennington I updated this PR to only delete the records of old datafile versions. The summary, datafile object, and parser errors are maintained.

jtimpe reviewed (approved):
Works well! I am still in favor of removing the `delay` on `remove_old_versions` - i was able to notice a delay when testing locally and the record counts appeared off until the new task ran. Other than that, looks good!

raftmsohani commented on tdrs-backend/tdpservice/scheduling/datafile_retention_tasks.py:
```diff
@@ -0,0 +1,100 @@
+"""Celery hook for datafile version cleanup task."""
+
+from __future__ import absolute_import
```
I think in newer version of python/django you don't need this anymore

elipe17 replied:
This is still a best practice as far as I know. Just make real sure there aren't any conflicts

raftmsohani replied:
But this is only required if we run code in pythonv2, in v3 absolute imports are added by default

mattcoleanderson reviewed (approved):
Works for me as well!

raftmsohani reviewed (approved):
I am still not sure we need absolute import here, but otherwise looks good to me

ADPennington committed:
Merge branch 'develop' into 3240-latest-df-only

elipe17 committed:
Merge branch 'develop' into 3240-latest-df-only

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3240-latest-df-only

elipe17 committed:
- Update so that only previous datafile versions records are deleted.

elipe17 committed:
- linting

ADPennington committed:
Merge branch 'develop' into 3240-latest-df-only

lhuxraft committed:
Merge branch 'develop' into 3240-latest-df-only

elipe17 committed:
- update to match new function def

elipe17 committed:
Merge branch 'develop' into 3240-latest-df-only

elipe17 committed:
Merge branch 'develop' into 3240-latest-df-only

elipe17 committed:
- Remove celery task
- Add correct logging to existing task
- Do not remove old versions of datafiles on submission
- Update DB task to run cleanup after backup success
- Logic change on log context
- Cleanup beat schedule

elipe17 commented:
@ADPennington I updated the Database backup task to run the clean up after it successfully performs it's backup. Unfortunately Celery Beat does not let you use future, past, or non-existent (i.e 2/31/XXXX) dates for crontabs. Because of this I can't add it to the admin console the way we discussed. It still has the management command and is tied to the backup task. Sorry for any inconvenience. If it becomes a real pain point we can write some custom code like we did for the reparse action to make it accessible from the DAC.

ADPennington committed:
Merge branch 'develop' into 3240-latest-df-only

ADPennington commented on tdrs-backend/tdpservice/scheduling/datafile_retention_tasks.py:
```diff
@@ -0,0 +1,58 @@
+                            continue
+                        newest_file = files.latest('version')
+                        ids = files.exclude(id=newest_file.id).values_list('id', flat=True)
+                        delete_records(ids, log_context)
```
@elipe17 this could be a follow-on, but i'm curious if we can get the following messages added to logentries as part of the records deletion process.

`logger.info(f"Deleting {count} records of type: {model}.")`

generally looking to get the total number of records in the search_indexes_... tables before and after the cleanup.

elipe17 replied:
I updated the function to log to DAC about the records it is deleting. Do note that this behavior will also occur for reparsing.

![Screenshot 2025-04-09 at 9 07 25 AM](https://github.com/user-attachments/assets/b9c11895-1a87-4b08-83a5-7f27ff8f5253)

First number is what is being deleted and the second number is the total number of records in the table. I also updated it to only log when it is actually deleting records.

ADPennington replied:
thanks @elipe17!  im okay with this for reparsing even though the reparsing model does a good job of tracking the before/after in the model. ready for the final test when this re-deploys

elipe17 committed:
- Add logging for record deletion

elipe17 committed:
- Update so that FRA is included

ADPennington reviewed (approved):
LGTM @elipe17 🚀 

the db cleanup task will be run nightly after the db backup task. some evidence of the records deleted in the logs below 😄 

cc: @klinkoberstar 


---

![Screenshot 2025-04-09 114509](https://github.com/user-attachments/assets/7bc71b2b-d90f-406d-a108-a400721f5e30)
![Screenshot 2025-04-09 114438](https://github.com/user-attachments/assets/d6d71aaf-3424-477f-9ba1-12e69a96a102)

elipe17 committed:
Merge branch 'develop' into 3240-latest-df-only
